pylabrobot/pumps/agrowpumps/agrowdosepump.py

```diff
@@ -0,0 +1,185 @@
+      if self.modbus.connected:
+        register_return = \
+          self.modbus.read_holding_registers(19, 2, unit=self.unit)
+        return int("".join(chr(r // 256) + chr(r % 256) for r in register_return.registers)[2])
+      else:
+        return int(0)
+    else:
+      return int(len(self.pump_index_to_address))
+
+  def start_keep_alive_thread(self):
+
+    """
+
+    keep_alive_thread(self): This method creates a daemon thread that sends a Modbus request
+    every 25 seconds to keep the connection alive.
+
+    """
+
+    def keep_alive():
+      """
+      keep_alive(self): This method sends a Modbus request every 25 seconds to keep the
+      connection alive.
+      """
+      assert self.modbus is not None, "Modbus connection not established"
+      while True:
+        time.sleep(25)
+
+        self.modbus.read_holding_registers(0, 1, unit=self.unit)
+
+    self.keep_alive_thread = threading.Thread(target=keep_alive, daemon=True)
+    self.keep_alive_thread.start()
+
+  async def setup(self):
+    """
+    setup(self): This method sets up the Modbus connection to the AgrowPumpArray and creates the
+    pump mappings needed to issue commands.
+
+    Awaitable:
+        self.modbus.connect(): This method connects to the AgrowPumpArray via Modbus.
+
+    """
+    if self.port == "simulated":
+      self.modbus = SimulatedModbusClient()
+    else:
+      self.modbus = ModbusClient(method="rtu", port=self.port,
+                                 baudrate=115200, timeout=1, stopbits=1, bytesize=8, parity="E",
+                                 retry_on_empty=True)
+    assert self.modbus is not None, "Modbus connection not established"
+    await self.modbus.connect()  # type: ignore
+    self.start_keep_alive_thread()
+    self.pump_index_to_address = {pump: pump + 100 for pump in range(self.num_channels)}
+
+  async def run_revolutions(self, num_revolutions: List[float],
+                            use_channels: List[int]):
+    """Run the specified channels at the speed selected. If speed is 0, the pump will be halted.
+      Args:
+        num_revolutions: number of revolutions to run pumps.
+        use_channels: pump array channels to run
+
+      Raises:
+        NotImplementedError: Revolution based pumping commands are not available for this array.
+    """
+    raise NotImplementedError(
+      "Revolution based pumping commands are not available for this pump array.")
+
+  async def run_continuously(self, speed: List[float],
+                             use_channels: List[int]):
+    """Run pumps at the specified speeds.
+        Args:
+            speed: rate at which to run pump.
+            use_channels: pump array channels to run
+
+        Raises:
+            ValueError: Pump address out of range
+            ValueError: Pump speed out of range
+    """
+    assert self.modbus is not None, "Modbus connection not established"
+    assert self.pump_index_to_address is not None, "Pump address mapping not established"
```

Comment:
Could these checks be put in a decorator?

Reply:
good thinking!